### ISSUE TITLE
Updating cilium configuration when endpoints are running for PolicyEnforcement causes deadlock

### ISSUE THREAD

**Issue by ianvernon** [cilium/cilium#1539]
Discovered by @eloycoto.

Reproducing steps:

* Run `cilium config PolicyEnforcement=always`
* Create some pods.
* Run `cilium config PolicyEnforcement=never`

Logs from a run of this: https://gist.github.com/eloycoto/b7ca626cf396c456eb71f90437031fa3

**Comment by ianvernon** [comment]
@eloycoto I have tried reproducing this manually on my local setup with the following commands at the root of the Cilium repository running in the K8s developer VM for cilium.

Steps I have taken:
1. Launch Cilium K8s developer VM from root if Cilium repository in my workspace:
`K8S=1 ./contrib/vagrant/start.sh`
2. SSH into the Cilium K8s developer VM:
`K8S=1 vagrant ssh`
3. With no pods running, configure policy enforcement mode to 'always'
`cilium config PolicyEnforcement=always`
4. Launch pods found in `examples/minikube/demo.yaml`:
`kubectl create -f examples/minikube/demo.yaml`
5. Configure PolicyEnforcement mode to 'never' after launching the pods:
`cilium config PolicyEnforcement=never`


Example output starting from step 3:
```
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ cilium config PolicyEnforcement=always
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ kubectl create -f examples/minikube/demo.yaml 
service "app1-service" created
deployment "app1" created
deployment "app2" created
deployment "app3" created
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ kubectl get pods
NAME                    READY     STATUS              RESTARTS   AGE
app1-3720119688-2pwmx   0/1       ContainerCreating   0          5s
app1-3720119688-xqwv2   0/1       ContainerCreating   0          5s
app2-1798985037-5m0v3   0/1       ContainerCreating   0          5s
app3-2097142386-1m4h6   0/1       ContainerCreating   0          5s
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ cilium endpoint list
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])                   IPv6                     IPv4            STATUS         
           ENFORCEMENT                                                                                                                    
173        Enabled       256        k8s:io.kubernetes.pod.namespace=kube-system   fd02::c0a8:210b:0:ad     10.11.42.252    regenerating   
                                    k8s:k8s-app=kube-dns                                                                                  
13949      Enabled       257        k8s:id=app1                                   fd02::c0a8:210b:0:367d   10.11.193.100   ready          
                                    k8s:io.kubernetes.pod.namespace=default                                                               
29381      Enabled       259        k8s:id=app3                                   fd02::c0a8:210b:0:72c5   10.11.13.37     regenerating   
                                    k8s:io.kubernetes.pod.namespace=default                                                               
48896      Enabled       257        k8s:id=app1                                   fd02::c0a8:210b:0:bf00   10.11.167.158   ready          
                                    k8s:io.kubernetes.pod.namespace=default                                                               
60670      Enabled       258        k8s:id=app2                                   fd02::c0a8:210b:0:ecfe   10.11.195.213   ready          
                                    k8s:io.kubernetes.pod.namespace=default                                                               
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ cilium endpoint list
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])                   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                                              
173        Enabled       256        k8s:io.kubernetes.pod.namespace=kube-system   fd02::c0a8:210b:0:ad     10.11.42.252    ready    
                                    k8s:k8s-app=kube-dns                                                                            
13949      Enabled       257        k8s:id=app1                                   fd02::c0a8:210b:0:367d   10.11.193.100   ready    
                                    k8s:io.kubernetes.pod.namespace=default                                                         
29381      Enabled       259        k8s:id=app3                                   fd02::c0a8:210b:0:72c5   10.11.13.37     ready    
                                    k8s:io.kubernetes.pod.namespace=default                                                         
48896      Enabled       257        k8s:id=app1                                   fd02::c0a8:210b:0:bf00   10.11.167.158   ready    
                                    k8s:io.kubernetes.pod.namespace=default                                                         
60670      Enabled       258        k8s:id=app2                                   fd02::c0a8:210b:0:ecfe   10.11.195.213   ready    
                                    k8s:io.kubernetes.pod.namespace=default                                                         
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ cilium config PolicyEnforcement=never
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ cilium endpoint list
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])                   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                                              
173        Disabled      256        k8s:io.kubernetes.pod.namespace=kube-system   fd02::c0a8:210b:0:ad     10.11.42.252    ready    
                                    k8s:k8s-app=kube-dns                                                                            
13949      Disabled      257        k8s:id=app1                                   fd02::c0a8:210b:0:367d   10.11.193.100   ready    
                                    k8s:io.kubernetes.pod.namespace=default                                                         
29381      Disabled      259        k8s:id=app3                                   fd02::c0a8:210b:0:72c5   10.11.13.37     ready    
                                    k8s:io.kubernetes.pod.namespace=default                                                         
48896      Disabled      257        k8s:id=app1                                   fd02::c0a8:210b:0:bf00   10.11.167.158   ready    
                                    k8s:io.kubernetes.pod.namespace=default                                                         
60670      Disabled      258        k8s:id=app2                                   fd02::c0a8:210b:0:ecfe   10.11.195.213   ready    
                                    k8s:io.kubernetes.pod.namespace=default                                                         
```

**Comment by ianvernon** [comment]
EDIT: I went ahead and made the script a bit smarter. 

Ah, now I am able to reproduce it with a script reliably:
```
#!/bin/bash

echo "setting PolicyEnforcement=default"
cilium config PolicyEnforcement=default
echo "cilium config output"
cilium config
echo "setting PolicyEnforcement=always"
cilium config PolicyEnforcement=always
echo "creating resources in examples/minikube/demo.yaml"
kubectl create -f examples/minikube/demo.yaml


NUM="5"
found=$(cilium endpoint list -o json | jq '. | length')
  while [[ "$found" -ne "$NUM" ]]; do
    if [[ $((iter++)) -gt $((5*60/1)) ]]; then
      echo ""
      exit 1
    else
      cilium endpoint list
      echo -n " [${found}/${NUM}]"
      sleep 1
    fi
    found=$(cilium endpoint list -o json| jq '. | length')
  done

echo "setting PolicyEnforcement=never"
cilium config PolicyEnforcement=never
echo "cilium config output"
cilium config
echo "listing endpoints"
cilium endpoint list
echo "deleting resources in examples/minikube/demo.yaml"
kubectl delete -f examples/minikube/demo.yaml
```

**Comment by ianvernon** [comment]
Still actively debugging this as I believe this is related to recent build failures. Added logs for when locks are released / acquired in this branch: 1539-policy-enforcement-deadlock

**Comment by ianvernon** [comment]
Finally nailed down the deadlock here, which occurs when endpoints are being regenerated and the value for `PolicyEnforcement` is changed via `cilium config`. In `endpoint/policy.go:regeneratePolicy`, the following happens:
```
        repo.Mutex.RLock()
        ...
	if owner.EnableEndpointPolicyEnforcement(e) {
		opts[OptionPolicy] = "enabled"
	} else {
		opts[OptionPolicy] = "disabled"
	}
        ...
	repo.Mutex.RUnlock()
```

`daemon/policy.go:EnableEndpointPolicyEnforcement` waits to take an RLock on `config.EnablePolicyMU`.

In `daemon/daemon.go`, at the same time, if `Handle(params PatchConfigParams)` is called while an endpoint has `config.EnablePolicyMU` held if a change to the configuration has been made:
```
...
	// Only recompile if configuration has changed.
	if changes > 0 {
		log.Debugf("daemon configuration has changed; recompiling base programs")
		d.conf.EnablePolicyMU.Lock()
		defer d.conf.EnablePolicyMU.Unlock()
		if err := d.compileBase(); err != nil {
			msg := fmt.Errorf("Unable to recompile base programs: %s", err)
			log.Warningf("%s", msg)
			return apierror.Error(PatchConfigFailureCode, msg)
		}
	}
...
```

`daemon/daemon.go:compileBase` calls `daemon/daemon.go:writeNetdevHeader`, which in turn calls `daemon/daemon.go:fmtPolicyEnforcement`. This function begins with the following:
```
d.GetPolicyRepository().Mutex.RLock()
enforcement := d.EnablePolicyEnforcement()
d.GetPolicyRepository().Mutex.RUnlock()
...
```

so, fmtPolicyEnforcement is waiting for the daemon's policy repository mutex to get released and its callers have Locked `EnablePolicyMU`, while `endpoint/policy.go:regeneratePolicy` is holding the policy repository mutex, and calls `daemon/policy.go:EnableEndpointPolicyEnforcement`, which is waiting for `EnablePolicyMU` to get unlocked.

The fix for this, funnily enough, is a PR that is approved and just needs me to add some docs: #1464. This PR removes the locking of the policy repository in `daemon/daemon.go:fmtPolicyEnforcement`,  as we do not need to access the policy repository in `daemon/policy.go: EnablePolicyEnforcement`.

**Comment by tgraf** [comment]
Fixed via #1464
